Add BTree and operations over it to ocean

nemanja-boric-sociomantic opened the pull request:
This PR introduces BTree data structure and algorithms over it to the ocean. This implementation of a BTree is templated on the type it carries, on the tree's degree, and on the allocation policy that should be used for the allocating the internal nodes (it is useful if the extending the BTree is done once in a while, but when we shouldn't give chance to GC to do mark&sweep). Additionally, GC tracking of the memory can be turned of, so the GC will not (somewhat expensively) traverse the tree, which is useful if we're storing value types, or the reference types for which we're storing the reference somewhere else.  In addition to basic functionality, inorder traversal of the tree, either via opApply or InputRange is supported.

nemanja-boric-sociomantic committed:
Add BTree and operations over it to ocean

This commits introduces BTree data structure and algorithms over it to
the ocean. This implementation of a BTree is templated on the type it
carries, on the tree's degree, and on the allocation policy that should
be used for the allocating the internal nodes (it is useful if the
extending the BTree is done once in a while, but when we shouldn't give
chance to GC to do mark&sweep). Additionally, GC tracking of the memory
can be turned of, so the GC will not (somewhat expensively) traverse the
tree, which is useful if we're storing value types, or the reference
types for which we're storing the reference somewhere else.

In addition to basic functionality, inorder traversal of the tree,
either via opApply or InputRange is supported.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/util/Allocators.d:
```diff
@@ -0,0 +1,265 @@
+/*******************************************************************************
```
This module looks similar to https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/util/container/mem/MemManager.d. Compare and contrast...

nemanja-boric-sociomantic replied:
Ah, I wasn't aware of this one. I'll give my best to use the existing one.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Internal.d:
```diff
@@ -0,0 +1,612 @@
+
+template BTreeImplementation()
+{
+    /// Inserts the element inside a leaf
```
Non-standard function doc. (Ditto for several other functions below.)

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Internal.d:
```diff
@@ -0,0 +1,612 @@
+    Internal operations over BTree.
+
+    Copyright:
+        Copyright (c) 2009-2017 Sociomantic Labs GmbH.
```
Copyright 2017. (Check all modules.)

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+/******************************************************************************
+
+    BTree structure.
```
Maybe move this comment block to the header and put `/// ditto` here?

gavin-norman-sociomantic replied:
Ah but then what about the `Params:` section... :(

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    mixin BTreeImplementation!() impl;
+}
+
+/*
```
The comments from here to line 294 are out-dated, or?

nemanja-boric-sociomantic replied:
Yeah, already force-pushed them.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Internal.d:
```diff
@@ -0,0 +1,612 @@
+import ocean.core.Enforce;
+import ocean.core.Traits;
+
+template BTreeImplementation()
```
What's the reason for splitting this into a separate module?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    - Every node has the following attributes:
+
+        x.number_of_elements = number of elements node contains
```
What does `x.` refer to here?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    - All leaves have the same depth
+
+    - Every node other than root must have at least (degree-1) elements.
```
Maybe explain what "degree" is before using the term?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+        x.number_of_elements = number of elements node contains
+        x.elements = the elements themselves
+        x.is_leaf = indicator if the node is the leaf.
+        x.child_nodes = number_of_elements+1 pointers to its children
```
Maybe explain previously the difference between elements and children?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+        x.is_leaf = indicator if the node is the leaf.
+        x.child_nodes = number_of_elements+1 pointers to its children
+
+    - Elements in the node split the ranges of the elements stored
```
"in a node" ? ("the node" implies there's only one, which sounds wrong.)

gavin-norman-sociomantic replied:
Generally I don't understand this sentence.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    - Every internal node other than the root has at least degree children.
+
+    - Every node may contain at most 2*degree - 1 keys. Therefore, an internal
```
"keys"?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+      node may have at most 2*degree children.
+
+    Other than insert and delete operations, inorder traversal method is
+    added, which traverses the tree in inorder order (left subtree is visited
```
"left subtree is visited first, then the element, then the right subtree" -- a couple of questions:
1. This makes it sound like each non-leaf node has exactly two children (the left subtree and the right subtree), which contradicts what is stated above.
2. "the element" makes it sound like each node has one single element. I guess it's a typo and you mean "the elements".

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    of the inorder traversal.
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
+        h is the tree height):
```
Extra indent.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+        - Searching O(h) = O(log_t(n))
+        - Inserting O(h) = O(log_t(n))
+        - Deleting: O(h) = O(log_t(n))
+
```
Would be nice to compare against other tree implementations in ocean. (There's at least the ebtree.)

gavin-norman-sociomantic replied:
An idea: maybe a README in `util.container.tree`?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    Params:
+        TreeElementType = type of the element to store.
+        tree_degree = degree of a tree (refer to the documentation)
```
"the tree"

gavin-norman-sociomantic replied:
`TrackingPolicy` not mentioned.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that the iteration is undefined
+        in case the tree is changed during the iteration.
```
Would be nice the expand on that a little. Will the iteration just stop? Will it continue but maybe miss elements? Is it possible elements could be iterated twice?

nemanja-boric-sociomantic replied:
I'll look to define this undefined behavior :+1:.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    {
+        if (this.root is null)
+        {
+            this.root = insertNewNode();
```
`this.insertNewNode()`?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    ***************************************************************************/
+
+    public Const!(ElementType)* insertElement (ElementType el)
```
Should `el` be const?

nemanja-boric-sociomantic replied:
Yeah... I see what I did here. In dlsnode, this structure used to contain only value types. This way it was very convenient when I would create a copy of the element to insert here. This approach doesn't work for the reference types.

nemanja-boric-sociomantic replied:
I thought about this one, and I think it needs to stay as it is. In case of value-types, this would create a copy of the value type, which we'll put into the tree. In case of the reference type, we will do the same - this will provide the copy of the reference and it would be put into the tree as well. My feeling about this being `const` is that it's too harsh - the object itself doesn't need to be const, _except_ when it comes to ordering - ordering needs to stay the same. Templating `BTree` on `Const!(Object)` would provide the guarantee to the user, if they wish. What do you think?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+     ******************************************************************************/
+
+    public bool removeElement (ref ElementType element)
```
Why are you passing the element by ref here but not to `insertElement`? Should it be const?

nemanja-boric-sociomantic replied:
Yeah, it should be const. What I wanted actually is pass `ref Const!(ElementType)`, but it didn't work in D1 for some reason (I'll reevaluate this), since I don't need a copy of the element, and I don't want to change it. I'll look into normalizing passing convention through the entire module.

nemanja-boric-sociomantic replied:
Ah, I know - it's the artifact from the time when this was a free-standing templated function, the template deduction wouldn't work.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    /******************************************************************************
+
+        Finds the highest value in the tree.
```
Highest/largest according to what definition?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    *******************************************************************************/
+
+    public Const!(ElementType)* search (ref ElementType element)
```
const?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    package struct BTreeNode
+    {
+        /// Number of the elements currently in the node
```
This contradicts the statement above "Contains degree * 2 - 1 elements". Should that say "at most"?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+******************************************************************************/
+
+struct BTree(TreeElementType, int tree_degree, alias Allocator = MallocAllocator,
```
What happens if `tree_degree` is negative or 0? :trollface:

nemanja-boric-sociomantic replied:
excellent question - I'm not sure :-) :+1:

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+        int number_of_elements;
+
+        /// Indicator if the given node is a leaf
+        bool is_leaf;
```
If this is true, are any conditions about the other fields implied? (invariant?)

nemanja-boric-sociomantic replied:
Yes, I will add the invariant for the `Node`.

gavin-norman-sociomantic replied:
Though... as it doesn't have any methods, it wouldn't be run...

nemanja-boric-sociomantic replied:
I think it should have some methods :-). I'll see about abstracting some of the elementary operations on the node to a node's methods.

gavin-norman-sociomantic replied:
Sounds good.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+        bool is_leaf;
+
+        /// Array of the elements
+        package TreeElementType[tree_degree * 2 - 1] elements;
```
Nice that these are fixed-size arrays :+1:

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    /// Convenience alias of the element type
+    package alias TreeElementType ElementType;
+
+    /// degree of a tree
```
"the tree"

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    /// degree of a tree
+    package const degree = tree_degree;
+
+    /// Root directory
```
node?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+    package BTreeNode* root;
+
+    /// "version" of the tree's state. Will be incremented on any removal/adding
+    /// the element. Used to check if the ranges are valid while doing the iteration
```
"doing an iteration"

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,695 @@
+
+    /// "version" of the tree's state. Will be incremented on any removal/adding
+    /// the element. Used to check if the ranges are valid while doing the iteration
+    package int content_version;
```
Is a signed integer large enough? If it's incremented on each modification, it wouldn't seem crazy that it could wrap under certain usage conditions.

nemanja-boric-sociomantic replied:
This is tricky. One one hand, we don't worry about the overflow here, since we are only interested in the difference between the current tree's version and the version which was there when the range was created. However, with this in place, it is possible that if you start a range and do 2^32 operations, while the range is still active, game over. I don't think this edge case should be allowed in a general purpose data structure. I'll see what I can do.

nemanja-boric-sociomantic replied:
Perhaps just using long (ulong really, we want defined wraparound) would be sufficient here. I'll see if I can make it cheap and defined, anyway.

gavin-norman-sociomantic reviewed (commented):
First high-level review focussing on doc, API, and some overall design questions.

gavin-norman-sociomantic commented:
>Move to ocean.util.container.mem.MemManager

Cool that this was possible!

nemanja-boric-sociomantic commented:
Squashed and ready for the next review round @gavin-norman-sociomantic .

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    BTree is a rooted tree, with the following properties:
+
+    - There's a characteristic of the tree, tree's degree, which indicates
```
Clearer: "A BTree has a characteristic called "degree". This indicates how many items are stored in each node of the tree. Having a high degree..."

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+      one go.
+
+
+    - The layout of one node of the tree is similar to the following:
```
"similar to" makes it sound like this is a trick and not actually the layout :D

"is as follows" ?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+      Every element in the node (for the non-leaf nodes) is surrounded by
+      two pointers. We call these pointers child nodes. Each pointer is a root
+      of a subtree, for which the following stands: all the elements in a tree
```
"for which the following holds true"

gavin-norman-sociomantic replied:
I think it would be clearer to move this to another "section":
```
- The layout of one node of the tree is as follows:
  -----------------------------------------------------------------
  | p1 | element1 | p2 | element2 | ... | pn-1 | element n-1 | pn |
  -----------------------------------------------------------------

  Every element in the node (for the non-leaf nodes) is surrounded by
  two pointers. We call these pointers child nodes.

- Ordering of elements: each child node pointer is the root of a subtree,
  for which the following holds true: all the elements in the tree ...
```

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    - Every node has the following attributes:
+
+        node.number_of_elements = number of elements node contains
```
Is this really stored or is it a constant? It sounds like this should == degree, which is a constant. Maybe the explanation of degree needs to make it clear that this is the _maximum_ number of elements in one node?

nemanja-boric-sociomantic replied:
This is actually stored. Remember that the we're allocating nodes as a static arrays. This keeps track of how many elements are _currently_ in the node. (Maximum is always specified as 2*degree-1).

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+        node.number_of_elements = number of elements node contains
+        node.elements = the elements themselves
+        node.is_leaf = indicator if the node is the leaf.
```
"a leaf"

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+        node.number_of_elements = number of elements node contains
+        node.elements = the elements themselves
+        node.is_leaf = indicator if the node is the leaf.
+        node.child_nodes = number_of_elements+1 pointers to it the subtrees whose
```
Typo "to it the"

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    - All leaves have the same depth
+
+    - Every node other than root must have at least (degree-1) elements.
```
And at most degree?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    - Every node other than root must have at least (degree-1) elements.
+
+    - Every internal node other than the root has at least degree children.
```
They can have more than degree children?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    - Every internal node other than the root has at least degree children.
+
+    - Every node may contain at most 2*degree - 1 elements. Therefore, an internal
```
Ah ok, this contradicts what was written above about degree...

nemanja-boric-sociomantic replied:
Ah, yes, of course, sorry for the confusion.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+    - Every node may contain at most 2*degree - 1 elements. Therefore, an internal
+      node may have at most 2*degree children.
+
+    Other than insert and delete operations, inorder traversal method is added,
```
"is added" is unusual wording. It implies that something has changed.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    /**************************************************************************
+
+        Constructor.
```
Params

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    /**************************************************************************
+
+        Inserts the element in the tree.
```
The element seems to be copied by value into the tree. It'd be nice to explicitly mention this, for user convenience.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+     ******************************************************************************/
+
+    public bool remove (ref ElementType element)
```
Would be useful to note why this is ref.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+        Finds the highest value in the tree.
+
+        Returns:
+            largest element in the tree, or ElementType.init if the tree is empty.
```
Using `ElementType.init` as the null value could be tricky with certain types. e.g. int: 0 is both a valid highest value and the null value.

nemanja-boric-sociomantic replied:
Ah, I think the API should be

```
bool findMaximum (out ElementType max)
```

nemanja-boric-sociomantic replied:
(there's one method, I think `search` where I return the indicator if the element was found. I think this is better way).

gavin-norman-sociomantic replied:
Yeah, sounds more generic.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+            found = indicator if the element was found
+
+        Returns:
+            pointer to the element, or ElementType.init if not found.
```
Sounds impossible.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+    ******************************************************************************/
+
+    public ElementType findMaximum ()
```
Why no `findMinimum` btw?

nemanja-boric-sociomantic replied:
I didn't need it. Will add it for completeness.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that, in case the tree is changed
+        during the iteration, iteration will not continue.
```
"iteration will halt" is a bit clearer.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+    }
+
+
+    alias BTreeImplementation!(ElementType, tree_degree)
```
Doc

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+    /// Private implementation
+    package Implementation impl;
+}
+
```
Maybe add a documented unittest / usage example here?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+{
+    BTree!(File, 2) tree;
+    tree.init();
+    assert (tree.findMaximum == File.init);
```
Any reason not to use `test`?

nemanja-boric-sociomantic replied:
Whoops. It's a leftover. I'll change all asserts here in test.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+
+module ocean.util.container.btree.BTree;
+
+import ocean.transition;
```
Do any of these need to be imported at module scope?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTree.d:
```diff
@@ -0,0 +1,507 @@
+import ocean.util.container.mem.MemManager;
+import ocean.util.container.btree.Implementation;
+
+public import ocean.util.container.btree.BTreeRange;
```
Probably would help to note the existence of this module in the doc (at a high level) somewhere. Otherwise, the user will only discover it by chance.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+
+/*******************************************************************************
+
+    Exception thrown from the BTreeRange if the iteration over the changed
```
"if a tree is modified during iteration" ?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+
+/*******************************************************************************
+
+    Returns a range that iterates the tree via the inorder.
```
"that performs an inorder iteration over the specified tree"

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    Returns a range that iterates the tree via the inorder.
+
+    If tree has changed during the iteration, range will throw an
+    RangeInvalidatedException. If the user knows that the tree will not change
```
I'm not sure the 2nd sentence says anything that the 1st hasn't already said.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    during the range traversal, it is safe to call range's methods - it will throw
+    an exception if this promise is broken. In case user knows that the tree
+    can be changed any time, it is possible to call BTreeRange.isValid before
+    performing any action on the range to check if the accessing is safe. This
```
What does "safe" mean here?

nemanja-boric-sociomantic replied:
I'll clarify more. It means that either fetching the current element, or moving to the next one is guaranteed to succeed in a single-thread environment.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+
+    Params:
+        tree = tree to iterate over
+        buff = buffer for storing the state of the iteration. Useful
```
Maybe specify what magnitude of data might be stored here. (A few bytes? A few Mb? A copy of the whole tree?)

nemanja-boric-sociomantic replied:
I will clarify. It just stores at most `tree_height * sizeof(pointer_to_a_node)` which should be really small, since tree height of this tree is small.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    /// Root node of a tree
+    private BTree.Implementation* tree;
+
+    /// "Version" of the tree at the iteration time.
```
"Copy of the content_version field of the tree at the time iteration began" ?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    private ulong tree_version;
+
+    /// Pair of the node/index for which we've paused iteration and pushed
+    /// to stack
```
The "for which" part isn't always true, I think. (See the `stack` field.)

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+        size_t index;
+    }
+
+    /// Stack of the nodes that we've postponed iteration over
```
Maybe add "used by ..." ?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    private struct IterationItem
+    {
+        BTree.Implementation.BTreeNode* node;
+        size_t index;
```
Index of what exactly?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    /// Stack of the nodes that we've postponed iteration over
+    private IterationItem[]* stack;
+
+    /// Current item we're on
```
Is there a more specific term than "item"?

gavin-norman-sociomantic replied:
Oh I see, "iteration item". I wonder whether the struct might be better named something like `NodeElement` or something that indicates what it's actually "pointing" to.

nemanja-boric-sociomantic replied:
Good one.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+    /// Current item we're on
+    private IterationItem item;
+
+    /// Current element over which we're iterating
```
Current element within the current `<something>`?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+
+    /***************************************************************************
+
+        Pops the next element from the tree.
```
"from the range"? (Not actually removing from the tree.)

nemanja-boric-sociomantic replied:
Correct, from the range, not from the tree.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+
+    ***************************************************************************/
+
+    private void ensureValid ()
```
`enforceValid`?

nemanja-boric-sociomantic replied:
This is now tricky. Because it doesn't _enforce_ anything, it just ensures that the range is valid before accessing it.

nemanja-boric-sociomantic replied:
I mean, it's not tricky, that was just my reasoning :-)

gavin-norman-sociomantic replied:
My reasoning was just that it does exactly what `enforce` does (checks a condition and throws, if it's false).

nemanja-boric-sociomantic replied:
You are correct, of course.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeRange.d:
```diff
@@ -0,0 +1,355 @@
+        }
+    }
+}
+
```
Documented unittest / usage example?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1121 @@
+
+    **************************************************************************/
+
+    package struct BTreeNode
```
Slightly more sensible to define this struct before using it (`root`).

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1121 @@
+
+    /***************************************************************************
+
+        Type of the element actually stored in the node. The type stored in the
```
I'd like D2 const masters to check this out.

@mihails-strasuns-sociomantic @mathias-lang-sociomantic -- casting away qualifiers inside a container... kosher or suspect?

nemanja-boric-sociomantic replied:
Yeah good idea. The only way around it I could think is that we recreate this array on every change.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1121 @@
+import ocean.util.container.mem.MemManager;
+
+
+/// ditto
```
Params :(

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1121 @@
+
+    /***************************************************************************
+
+        Constructor's like initialization memory. Prepares the tree to be used.
```
Comprehensibility of the first sentence is low ;)

nemanja-boric-sociomantic replied:
haha, I've just noticed this. Epic sentence.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1121 @@
+
+    ******************************************************************************/
+
+    private bool deleteFromNode ( BTreeNode* node, BTreeNode* parent,
```
I notice a lot of methods that accept a `BTreeNode*` as their first argument. Would it make sense to place these inside `BTreeNode`?

gavin-norman-sociomantic replied:
I also wonder about the pros and cons of having `BTreeNode` nested. It seems the only real reason is to share the template args of the outer struct. A pro of moving it outside would be (perhaps) increased code readability.

nemanja-boric-sociomantic replied:
Might be a good idea, I will evaluate this carefuly.

nemanja-boric-sociomantic replied:
Ok, I've looked into this and I think these methods should stay outside the structure. Reason is that it looks much more natural to look into the implementation, knowing these methods are using recursion to do work, when you're explicitly passing the node on which they are going to work.

gavin-norman-sociomantic replied:
I've lost track of this a bit. `deleteFromNode` is now _inside_ a struct  (`BTreeMapImplementation`). I'm not sure if that was the one I was referring to originally. This seems to contradict your comment:
>I've looked into this and I think these methods should stay outside the structure

nemanja-boric-sociomantic replied:
Sorry, I was referring to the fact that these methods are not nested inside the struct they are operating on (`BTreeNode`). This is just a style issue, and tbh, it feels more natural to define the methods that are operating recursively over the data structure as to have their node explicitly passed as the first argument. If you look into `deleteFromNode` it starts from the node that's given (usually we start with `root` and then it either finds the element in the node, or descends into one of the subtrees using recursion `return deleteFromNode(subtree_root, element)`. I feel this is much more natural (and I'm biased because this is the usually how the algorithms over the data structure are represented) than `subtree_root.deleteFromNode(element)`.

gavin-norman-sociomantic reviewed (commented):
2nd pass. Getting into mid-level stuff and touching on a few broad topics re: the lowest level.

nemanja-boric-sociomantic commented:
Thanks for the great comments @gavin-norman-sociomantic !

nemanja-boric-sociomantic commented:
@gavin-norman-sociomantic I've squashed the changes, so it's ready again.

The most important change is that I moved from the set to the map. This makes the API more clear and akin to other containers in `ocean.util.containers.*`. I don't think adding set would be much work, but I would leave that for latter. Some renamings are done, and the iteration now iterates over the (key, value) pairs.

mihails-strasuns-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+        }
+        else
+        {
+            int opCmp(typeof(this) rhs)
```
coverage check shows this method is not used

mihails-strasuns-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+import ocean.transition;
+import ocean.util.container.mem.MemManager;
+
+public import ocean.util.container.btree.BTreeMapRange;
```
I am not a fan of public imports in non-`package.d` modules. Explicitly importing range style API when you need it fits our code style better.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    - Each pointer is a root of a subtree, for which the following holds true: all
+      the elements in a tree pointed by p(i-1) (pointer left of the elements) are
+      smaller (in respect to the ordering defined by the element's type) of the
```
Does this discussion of ordering need to be updated now to account for the presence of keys?

nemanja-boric-sociomantic replied:
Yes, you're right. I should mention this now.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+        node.number_of_elements = number of elements node currently contains
+            (as nodes are statically allocated block, the number of the elements
+             may not necessarily be maximal.
```
Closing ) missing.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+    There's a recursive and range-based implementation of the inorder
+    traversal.
+
+    Range-based traversal is supported through `BTreeMapRange` structure, found in
```
"found in ocean.util.container.btree.Btree and publicly imported here."

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+      | p1 | element1 | p2 | element2 | ... | pn-1 | element n-1 | pn |
+      -----------------------------------------------------------------
+
+      Every element in the node (for the non-leaf nodes) is surrounded by
```
Maybe mention here that each element consists of a key and a value.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1259 @@
+
+    ***************************************************************************/
+
+    package IMemManager allocator;
```
Should probably `verify` that this is non-null before using it. (It's possible to use an instance without calling `init`.)

gavin-norman-sociomantic replied:
An alternative would be to remove `init`, rely on this field being set when constructing an instance, and add an invariant that checks it's non-null.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1259 @@
+
+    package void init (IMemManager allocator = mallocMemManager)
+    {
+        assert (this.allocator is null);
```
`verify`

nemanja-boric-sociomantic replied:
Ha! :-)

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+        the reference.
+
+        Params:
+            el = element to insert.
```
Check params doc for all methods that deal with keys now.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+    /******************************************************************************
+
+        Deletes the element from the BTreeMap. The element is passed by reference,
+        since we'll just use it for the comparison with the value stored in the tree.
```
No longer true.

nemanja-boric-sociomantic replied:
Oh, I'm going to do a documentation scan for the key/value.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+        Finds the highest value in the tree.
+
+        Params:
+            el = out parameter to be filled with the copy of the highest element
```
Params/returns backwards. Ditto for `findMinimum`. What was the reasoning for switching these?

gavin-norman-sociomantic replied:
Interesting. This searches for the largest value.
1. Is this useful? The key determines the ordering, so I would have thought the largest key would be the expected function.
2. Whatever the search method, the user will require the key & value returned.

nemanja-boric-sociomantic replied:
Yes, this is more useful for the set, it's not very useful now. I'll remove it, since the ordering is done by keys and this would return the item with the highest key.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that, in case the tree is changed
```
Please add "over keys and values".

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    /***********************************************************************
+
+        Recursive inorder iteration. Note that, in case the tree is changed
```
Please add "over values only".

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+        ElementType = type of the element to store.
+        tree_degree = degree of the tree (refer to the documentation)
+        memManager = memory management strategy to use
+
```
Returns

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+}
+
+
+///
```
I wonder if this should be a usage example for the struct, not for the constructor function.

nemanja-boric-sociomantic replied:
Yes, but I would highly discourage using this struct without constructor function. I think that can be doable with removing `init` function and adding invariant.

mihails-strasuns-sociomantic replied:
You can `@disable this` for actual struct and prepare it in constructor function from `T.init` state.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    *******************************************************************************/
+
+    public ElementType search (KeyType key, out bool found_element)
```
Maybe `get` would be a better name. `search` has a slight implication of just checking whether something exists, not necessarily returning its value.

nemanja-boric-sociomantic replied:
Very nice point, I'll do it.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+}
+
+// shared between different ranges
+private static RangeInvalidatedException range_exception;
```
`static` is meaningless here.

nemanja-boric-sociomantic replied:
Sorry, C habits die hard.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+
+    Params:
+        tree = tree to iterate over
+        buff = buffer for storing the state of the iteration. Useful
```
Mention behaviour if null.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+    }
+
+    /// Stack of the nodes that we've postponed iteration over
+    private NodeElement[]* stack;
```
Could an invariant check that this is never null?

gavin-norman-sociomantic replied:
Maybe expand this a bit. Something like: "When interationStep() moves from a node to iterate over its children, the parent node is pushed to the stack.".

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+    }
+
+    /// Stack of the nodes that we've postponed iteration over
+    private NodeElement[]* stack;
```
Invariant checking that this is never null?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+    /// Current (node, element_index) pair we're on
+    private NodeElement item;
+
+    /// Current Element (item.node[item.element_index].value)
```
"Current value"

gavin-norman-sociomantic replied:
Would it make sense to make `current_element` and `current_key` methods of `NodeElement`? (i.e. `key` and `value`.)

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    Params:
+        KeyType = type of the key.
+        TreeElementType = type of the element to store.
```
The terminology around values / elements could be tidied up. I'd suggest that an "element" is a key/value pair. Otherwise always refer to values (not elements).

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+
+    ***************************************************************************/
+
+    public Pair!(BTreeMap.KeyType, BTreeMap.ElementType) front ()
```
`alias Pair!(BTreeMap.KeyType, BTreeMap.ElementType) KeyValue` (or something)?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+        - Thomas H. Cormen et al. "Introduction to Algorithms, 3rd edition"
+
+    Params:
+        KeyType = type of the key.
```
`TreeKeyType`

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,388 @@
+
+    ***************************************************************************/
+
+    private void iterationStep ()
```
It occurs to me that the logic in this method -- moving from one node-element to the next, in iteration order -- must be duplicated somewhere inside the foreach functions of the tree implementation. Is there any sensible way this could be abstracted?

nemanja-boric-sociomantic replied:
I'll look into this. The logic inside the tree uses the call-stack for the recursion, while this allows us to suspend the iteration and resume it later (unless the map has changed in the meantime).

gavin-norman-sociomantic reviewed (commented):
Next pass. Looking into the range.

mihails-strasuns-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+    Range-based traversal is supported through `BTreeMapRange` structure, found in
+    (and here publicly imported) in ocean.util.container.btree.Btree.
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
```
More useful is to state complexity for reach public method in its own DDOC

mihails-strasuns-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+    ***************************************************************************/
+
+    package void init (IMemManager allocator = mallocMemManager)
```
It is not a good idea to name method `init` because of paren-less calls and built-in `T.init` defined for all types.

mihails-strasuns-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,664 @@
+
+******************************************************************************/
+
+public BTreeMap!(KeyType, ElementType, tree_degree) createBTreeMap
```
Phobos uses `makeX` notation (but we don't need to care ;))

mihails-strasuns-sociomantic reviewed (commented):
Some very brief comments without reading into implementation/tests

leandro-lucarella-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,570 @@
+    B-Tree data structure and operations on it.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
```
I would go with the updated lowercase copyright notice in new files :)

nemanja-boric-sociomantic commented:
Updated with feedback from all of you.

mihails-strasuns-sociomantic reviewed (approved):
@gavin-norman-sociomantic do you want to give another review pass? To me API looks reasonable enough to merge and everything else can be fixed later if needed.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,590 @@
+        number_tree.insert(i, i);
+    }
+
+    // let's find 7_001
```
What's 7,001?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMap.d:
```diff
@@ -0,0 +1,590 @@
+}
+
+unittest
+{
```
Might be nice to break this down into separate unittests, to make it easier to follow. There's at very least the test for the file map and the tests for the integer map.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/BTreeMapRange.d:
```diff
@@ -0,0 +1,412 @@
+        test!("==")(res, element);
+    }
+
+    // reseed, so that the difference two random number generated sets
```
Where does it reseed?

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1200 @@
+        private struct KeyValue
+        {
+            KeyType key;
+            Unqual!(ValueType) value;
```
Is the use of `Unqual` here described anywhere? This seems like an important implementation detail to mention at a high level.

gavin-norman-sociomantic commented on src/ocean/util/container/btree/Implementation.d:
```diff
@@ -0,0 +1,1200 @@
+
+    ******************************************************************************/
+
+    package int inorder (int delegate(ref KeyType key, ref ValueType value) dg)
```
Might be nice to combine the two `inorder` overloads. A template with a `static if` to call the different delegate types?

nemanja-boric-sociomantic commented:
Updated to address @gavin-norman-sociomantic 's comments (mostly outdated comments, separate unittests, de-duplication of `inorder`, commend on unqualed value type).

mihails-strasuns-sociomantic commented:
Ok, let's do it :P
